### PR TITLE
Add logic to broadcast current scheduler info to workers in CLI

### PR DESCRIPTION
**Pull request by lastephey**
This PR addresses the problem described in @rcthomas issue #45 

I think there was some confusion in this issue-- at NERSC our idea at the moment is to recommend that our users start Dask via the dask-mpi CLI (which never touches core.py, as far as we can tell.)

This PR implements the same logic in cli.py that already exists in core.py. It first starts the scheduler and then broadcasts the scheduler address to the workers before they can start. This behavior prevents the workers from reading an old scheduler.json file (which may have been left behind after a non-clean exit) and then hanging while they try to connect to a scheduler which does not exist.

### CODE REVIEW

**Comment by lastephey** [comment]
Clearly I should have tested this locally first. :) I'll get a test setup going to fix these errors.

**Comment by mrocklin** [comment]
Hi @lastephey !  Thanks for doing this.  In principle this looks good to me, and seems to resolve the problem.  

I suspect that @kmpaul has thoughts though, and he may also be able to shed some light on the CI failures.  I think that he only works M-Th though, so we may not hear back from him for the next day or two.

**Comment by lastephey** [comment]
Hi @mrocklin , thanks for the info! I'll keep thinking about this and hopefully @kmpaul has some insight. One issue is that I don't have a great setup for testing-- we can't use mpirun on Cori. Sorry for the noise from using your web CI for troubleshooting.

**Comment by mrocklin** [comment]
I conda install openmpi or mpich on my personal linux machine for testing
dask-mpi

**Comment by lastephey** [comment]
Thanks for the advice, I'll do that.

**Comment by kmpaul** [comment]
@lastephey Thanks for looking at this!  As @mrocklin pointed out, I have a short work schedule, so I couldn't respond until today.  

@mrocklin is correct.  The best way to test is to do it via a conda install with `mpi4py` on your laptop.  Install `mpich` and/or `openmpi` for testing.

**Comment by kmpaul** [comment]
So, these kinds of errors have come up quite frequently when running with MPI.  The work-around has always been to disable the `--nanny` option.  (That is, use the `--no-nanny` option.)  If you do this, then the tests pass.

I'm going to look a little further to see if I can figure out what is happening.  I've always suspected that MPI does not like it when you fork/spawn new processes from a running MPI process.  If that is the case, then I think that maybe we need to always use the `--no-nanny` option (or make it the default).

**Comment by lastephey** [comment]
Thanks for looking into this @kmpaul . Oddly all the tests pass for me on my laptop using conda installed `mpi4py` with both `--nanny` or `--no-nanny`. I am packing up my setup in a container in case it's useful.

**Comment by kmpaul** [comment]
Interesting!  I'm able to reproduce the errors both on my laptop and in a docker container.  Have you checked out your development branch?

**Comment by lastephey** [comment]
Yes. I checked out the `fix_mpi_sched` branch from my forked dask-mpi repo. 

What kind of mpi are you using/how did you install it? 

Screen Shot 2019-12-16 at 11 05 48 AM

**Comment by kmpaul** [comment]
Everything is installed from the `.circleci/install.sh` script.  You should just be able to do the following:

```bash
$ docker run -it continuumio/miniconda:latest
```

And then from the container run:

```bash
$ git clone https://github.com/lastephey/dask-mpi
$ cd dask-mpi
$ git checkout fix_mpi_sched
$ export PYTHON=3.6
$ export MPI=mpich
$ export ENV_NAME=dask-mpi-dev
$ .circleci/install.sh
$ pytest --verbose dask_mpi/tests/
```

**Comment by lastephey** [comment]
Thanks @kmpaul . I followed your directions exactly except I needed to additionally `pip install pytest` inside the container and `conda activate dask-mpi-dev`. I can reproduce the hang that happens during the CI testing and that you also see. 

But... this change in `cli.py` seems to be working in practice, it just fails the unit tests. Is there some way to make the tests more flexible/robust?

**Comment by kmpaul** [comment]
Looking at the `test_core.py` and `test_cli.py` tests, you will notice that the `test_core.py` tests use `subprocess.Popen` to launch the MPI job, while the `test_cli.py` tests use `distributed.utils_test.popen` to launch the MPI job (and the `distributed.utils_test.loop` fixture).  I'm not sure if one of these could be the problem.

**Comment by kmpaul** [comment]
Well, I've tried playing with these fixtures a bit but to no avail.  Still hanging.  I'm going to look more at this tomorrow.

**Comment by kmpaul** [comment]
@lastephey Just so you know, I've had problems with the testing framework in the past.  It has been challenging to test without problems.  I believe you that the implementation in `cli.py` works.  Now the challenge is to get tests that accurate demonstrate it.

**Comment by lastephey** [comment]
@kmpaul Thanks for your help with this. I'll try poking at it as well.

**Comment by kmpaul** [comment]
@lastephey I believe I know what is happening.

If you look at the `dask_mpi/tests/test_cli.py` tests, you will notice that they assume the use of a `scheduler_file` parameter when initializing the `Client`.

https://github.com/dask/dask-mpi/blob/de00068c22d735fca6a1aa561fc176f99ac3385d/dask_mpi/tests/test_cli.py#L24-L37

What we want to do is change line 31 (of  `dask_mpi/tests/test_cli.py`) from initializing the client via `Client(scheduler_file=fn)` to something like `Client(address=URL)`...but the problem is you don't actually know the `URL` from within the test.  And I can't think of a way of guaranteeing what the URL should be in the test itself.

@mrocklin Perhaps you have some ideas on this?

File: dask_mpi/tests/test_cli.py [hunk 1]
```diff
@@ -25,7 +25,7 @@
 def test_basic(loop, nanny, mpirun):
     with tmpfile(extension="json") as fn:
 
-        cmd = mpirun + ["-np", "4", "dask-mpi", "--scheduler-file", fn, nanny]
+        cmd = mpirun + ["-np", "4", "dask-mpi", "--scheduler-file", fn, nanny, "&"]
 
         with popen(cmd):
             with Client(scheduler_file=fn) as c:
```
Comment (kmpaul) [inline dask_mpi/tests/test_cli.py]: This is the culprit line causing the tests to hang.  It is in this line that the `distributed.Client` is being initialized and told to find the Scheduler by looking up the `address` parameter in the `scheduler_file`.  The `scheduler_file` (called `fn` in this test) is not being used by `dask-mpi`, so no information is being written to the `scheduler_file` from the Scheduler.  Hence, the Client keeps waiting and doesn't know how to find and connect to the Scheduler.

It is unclear to me how this should change if the Scheduler uses an MPI broadcast to tell the Workers to what address to connect.  The Client is left out of the broadcast, and so has no means of knowing where the Scheduler is located.  (This is different in the `core.py` code, where the Client's process actually participates in the broadcast, and therefore knows where to the Scheduler is located.)

The correct solution to this problem probably requires some discussion.

Comment (kmpaul) [reply dask_mpi/tests/test_cli.py]: I might also note that when you run `dask-mpi` by hand, as is commonly done, then you can get the address of the Scheduler from the log messages, and you can connect your Client to that.  In this case, though, when you are automating the launch of `dask-mpi` in the test, the only way I know of getting the Scheduler address is (1) through a `scheduler_file` or (2) scrubbing STDOUT (or the log file) from the Scheduler process to find the address.  Option (1) defeats the purpose of using the MPI broadcast, and Option (2) seems messy to me.

File: dask_mpi/tests/test_cli.py [hunk 2]
```diff
@@ -40,7 +40,7 @@ def test_basic(loop, nanny, mpirun):
 def test_no_scheduler(loop, mpirun):
     with tmpfile(extension="json") as fn:
 
-        cmd = mpirun + ["-np", "2", "dask-mpi", "--scheduler-file", fn]
+        cmd = mpirun + ["-np", "2", "dask-mpi", "--scheduler-file", fn, "&"]
 
         with popen(cmd, stdin=FNULL):
             with Client(scheduler_file=fn) as c:
```
Comment (kmpaul) [inline dask_mpi/tests/test_cli.py]: Again, this will only work if the Scheduler uses the `scheduler_file`.

File: dask_mpi/cli.py [hunk 3]
```diff
@@ -72,24 +72,32 @@ def main(
...
-                port=scheduler_port,
-            ) as s:
-                await s.finished()
+            ) as scheduler:
+                comm.bcast(scheduler.address, root=0)
+                comm.Barrier()
+                await scheduler.finished()
+
+        asyncio.get_event_loop().run_until_complete(run_scheduler())
+        sys.exit()
```
[the remark this answers is not in the document]
Comment (lastephey) [reply dask_mpi/cli.py]: i'll remove both of these

File: dask_mpi/cli.py [hunk 4]
```diff
@@ -1,13 +1,13 @@
 import click
 
 import asyncio
-from mpi4py import MPI
-from dask.distributed import Scheduler, Worker, Nanny
-from distributed.cli.utils import check_python_3
+import sys
```
[the remark this answers is not in the document]
Comment (lastephey) [reply dask_mpi/cli.py]: i'll remove this

File: dask_mpi/tests/test_cli.py [hunk 5]
```diff
@@ -25,7 +25,7 @@
 def test_basic(loop, nanny, mpirun):
     with tmpfile(extension="json") as fn:
 
-        cmd = mpirun + ["-np", "4", "dask-mpi", "--scheduler-file", fn, nanny]
+        cmd = mpirun + ["-np", "4", "dask-mpi", "--scheduler-file", fn, nanny, "&"]
```
[the remark this answers is not in the document]
Comment (lastephey) [reply dask_mpi/tests/test_cli.py]: ok sure, sorry for my misunderstanding. i'll remove all of these.

File: dask_mpi/cli.py [hunk 6]
```diff
@@ -72,24 +72,32 @@ def main(
...
+    rank = comm.Get_rank()
+
     if rank == 0 and scheduler:
 
-        async def run():
+        async def run_scheduler():
             async with Scheduler(
                 interface=interface,
                 protocol=protocol,
-                scheduler_file=scheduler_file,
```
[the remark this answers is not in the document]
Comment (lastephey) [reply dask_mpi/cli.py]: sure, i'll add this back in

File: dask_mpi/tests/test_cli.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (kmpaul) [reply dask_mpi/tests/test_cli.py]: No worries.  😄

File: dask_mpi/cli.py [hunk 7]
```diff
@@ -72,24 +72,32 @@ def main(
...
     else:
+        scheduler_address = comm.bcast(None, root=0)
+        dask.config.set(scheduler_address=scheduler_address)
+        comm.Barrier()
 
-        async def run():
+        async def run_worker():
             WorkerType = Nanny if nanny else Worker
             async with WorkerType(
-                scheduler_file=scheduler_file,
```
[the remark this answers is not in the document]
Comment (lastephey) [reply dask_mpi/cli.py]: added back in

File: dask_mpi/cli.py [hunk 8]
```diff
@@ -72,24 +72,32 @@ def main(
...
+        async def run_scheduler():
             async with Scheduler(
                 interface=interface,
                 protocol=protocol,
-                scheduler_file=scheduler_file,
                 dashboard_address=dashboard_address,
-                port=scheduler_port,
-            ) as s:
-                await s.finished()
+            ) as scheduler:
```
[the remark this answers is not in the document]
Comment (lastephey) [reply dask_mpi/cli.py]: changed `scheduler` to `s` in this instance

File: dask_mpi/cli.py [hunk 9]
```diff
@@ -72,24 +72,32 @@ def main(
...
+                comm.bcast(scheduler.address, root=0)
+                comm.Barrier()
```
[the remark this answers is not in the document]
Comment (lastephey) [reply dask_mpi/cli.py]: ok added

File: dask_mpi/cli.py [hunk 10]
```diff
@@ -72,24 +72,32 @@ def main(
...
+                await scheduler.finished()
+
+        asyncio.get_event_loop().run_until_complete(run_scheduler())
+        sys.exit()
 
     else:
+        scheduler_address = comm.bcast(None, root=0)
+        dask.config.set(scheduler_address=scheduler_address)
+        comm.Barrier()
```
[the remark this answers is not in the document]
Comment (lastephey) [reply dask_mpi/cli.py]: added

File: dask_mpi/cli.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (lastephey) [reply dask_mpi/cli.py]: i added an `if not scheduler_file` just after `as s` and it causes the scheduler to close and the program to hang. i believe we want to send the scheduler info no matter what, so i'm going to leave this out.

File: dask_mpi/cli.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (lastephey) [reply dask_mpi/cli.py]: i also added this `if not scheduler_file` and it causes the program to hang. i believe we want to send the scheduler info no matter what, so i'm going to leave this out.

File: dask_mpi/tests/test_cli.py (hunk 1 above)
Comment (lastephey) [reply dask_mpi/tests/test_cli.py]: Clearly this decision is up to the main developers, but IMHO the unit tests should reflect how people are commonly using `dask-mpi` and not the corner cases that stem from the CI setup.

That said, I very much appreciate your help in carefully looking over this PR @kmpaul and @mrocklin !

Comment (kmpaul) [reply dask_mpi/tests/test_cli.py]: You are more than welcome, @lastephey!  And I very much appreciate your PR.  Help on OSS is always appreciated.  😄 

I wouldn't call this a corner case.  The difficulty in getting this test to pass with the previous code changes was a sign that something was wrong.  And the issue was that by using an MPI broadcast to send the scheduler info to the workers, there was no way to *also* send the scheduler info to the *client* (i.e., the client created in the test).  Not without using a `scheduler_file`, and the use of a `scheduler_file` seems redundant if you are using an MPI broadcast.

**Comment by lastephey** [comment]
Hi @kmpaul 

I think we are approaching this problem from two different viewpoints. I think you are viewing the client as a part of the cluster (hence assigning it to rank=1 in core.py) and we would prefer to run the client separately for ease of viewing the dashboard on our shared system. 

This is why we wanted to use MPI to sync the scheduler info between the scheduler and the workers, but continue to use the `scheduler_file` for the client (which we instantiate separately from the cluster). 

To give you an idea of how we are using dask-mpi at NERSC at the moment, you can check out some of @rcthomas notebooks: https://gitlab.com/NERSC/nersc-notebooks/tree/master/dask

Does this help make our motivation more clear? Of course if there is a better way to make our setup work, we are open to suggestions.

**Comment by kmpaul** [comment]
@lastephey Ok.  I'm very sorry for missing the reason for wanting both a `scheduler_file` *and* MPI broadcasting.  I'm finally coming around.

However, I still don't think you actually need to MPI broadcast.  I think you just need some appropriately placed MPI Barriers to make sure that the Scheduler has started (and wrote the new `scheduler_file`) *before* the Workers start (and attempt to read the `scheduler_file`).  Doesn't that seem like it should fix your problem?

**Comment by kmpaul** [comment]
If you could accomplish this with just MPI Barriers, then it might be nice to add a test demonstrating the problems you experience with lingering stale `scheduler_file`s.

**Comment by lastephey** [comment]
@kmpaul ok good, I'm glad we are on the same page. We should have been more clear about our setup-- sorry for the confusion. 

Yes I think appropriately placed barriers would do just as well. Let me try to get that working and I'll add a stale scheduler file unit test, too.

**Comment by kmpaul** [comment]
Cool!  Again, I'm really sorry for all the confusion.  I feel like I just asked you to do a bunch of work that may not be necessary at all.  Sorry.  I would hate to have done that.

My first thoughts might be to just remove the `dask.config` and `comm.bcast(...)` lines entirely, leaving just the `comm.Barrier()` lines.  That might do the trick.  And if that's still not enough, you might add an `asyncio.sleep(0.1)` or something just before the Scheduler rank's barrier call to allow for enough time to write the new `scheduler_file`.

Tricky problem, but I think we might be narrowing on the solution.  😄

**Comment by lastephey** [comment]
The barrier solution seems to work ok even without the asyncio sleep, so i left it out.

I'll work on the unit test now.

**Comment by lastephey** [comment]
I have no idea why the tests failed before and passed now. I didn't make any meaningful changes.

**Comment by kmpaul** [comment]
Yeah.  That's very weird, but I've seen instabilities in MPI CI testing before.  I've never fully understood why this happens.

**Comment by lastephey** [comment]
ok thanks @kmpaul ! do you want the new unit test as a new PR?

**Comment by kmpaul** [comment]
Ach!  Sorry.  I was trigger happy.  

Yes.  Another PR would be great.

**Comment by lastephey** [comment]
Hi @kmpaul, just wondering if you're planning a release any time soon? I see the last release was Oct 2019 and this patch went in Jan 2020. It would be great to have a release install of installing from a commit.

**Comment by kmpaul** [comment]
@lastephey:  This is now done.  Version 2.21.0 of dask-mpi is released on PyPI and conda-forge.

**Comment by kmpaul** [comment]
...though the conda-forge release may take a bit to propagate out.  Should be available for download soon.

**Comment by lastephey** [comment]
Thank you very much for the fast turnaround!

**Comment by kmpaul** [comment]
You're welcome.  😄 I wish I could be this fast all the time.
